Wikipedia/Code/ArticleViewController+LinkPreviewing.swift

```diff
@@ -1,28 +1,151 @@
+                        didCallCompletion = true
+                    }
+                }
+            } else {
+                if (!didCallCompletion) {
+                    completionHandler(.success, config)
+                    didCallCompletion = true
+                }
+            }
+        }
+    }
+
     func getPeekViewControllerAsync(for linkURL: URL, completion: @escaping (UIViewController?) -> Void) {
         let destination = configuration.router.destination(for: linkURL)
         getPeekViewControllerAsync(for: destination, completion: completion)
     }
-    
+
+    func previewMenuElements(for previewViewController: UIViewController, suggestedActions: [UIMenuElement]) -> UIMenu? {
+        guard let vc = previewViewController as? ArticleViewController else {
+            return nil
+        }
+        let legacyActions = vc.previewActions
+        let menuItems = legacyActions.map { (legacyAction) -> UIMenuElement in
+            return UIAction(title: legacyAction.title) { (action) in
+                legacyAction.handler(legacyAction, previewViewController)
+            }
+        }
+        return UIMenu(title: "", image: nil, identifier: nil, options: [], children: menuItems)
+    }
+}
+
+// MARK: Peek/Pop for Protocol (iOS 12 and earlier, on devices w/ 3D Touch)
+// All functions in this extension are for 3D Touch menus. (Can be removed when the oldest supported version is iOS 13.)
+extension ArticleContextMenuPresenting {
     func getPeekViewController(for linkURL: URL) -> UIViewController? {
         let destination = configuration.router.destination(for: linkURL)
         return getPeekViewController(for: destination)
     }
-    
-    func getPeekViewController(for destination: Router.Destination) -> UIViewController? {
-        switch destination {
-        case .article(let articleURL):
-            let articleVC = ArticleViewController(articleURL: articleURL, dataStore: dataStore, theme: theme)
-            articleVC?.articlePreviewingDelegate = self
-            articleVC?.wmf_addPeekableChildViewController(for: articleURL, dataStore: dataStore, theme: theme)
-            return articleVC
-        default:
+
+    func shouldPreview(linkURL: URL?) -> Bool {
+        guard let linkURL = linkURL else {
+            return false
+        }
+        return linkURL.isPreviewable
+    }
+
+    func previewingViewController(for linkURL: URL?) -> UIViewController?  {
+        guard let linkURL = linkURL, let peekVC = getPeekViewController(for: linkURL) else {
             return nil
         }
+        hideFindInPage(nil)
+        return peekVC
     }
-    
+}
+
+// MARK:- Context Menu for ArticleVC (iOS 13 and later)
+// All functions in this extension are for Context Menus (used in iOS 13 and later)
+extension ArticleViewController: ArticleContextMenuPresenting, WKUIDelegate {
+    static let peekableImageExtensions: Set<String> = ["jpg", "jpeg", "gif", "png", "svg"]
```

Comment:
If @joewalsh has a reason to keep it, I can add it back (here or in another PR). For now, it's removed - and this should be ready for another review.

Reply:
It can be removed - I brought it over when converting the old ArticleVC but realized after the fact that it should be updated to utilize the new media list and router to determine peek destination.